core/fetcher.go

```diff
@@ -160,64 +171,82 @@ func (f *BlockFetcher) ValidatorSet(ctx context.Context, height int64) (*types.V
 	return validatorSet, nil
 }
 
-// SubscribeNewBlockEvent subscribes to new block events from Core, returning
-// a new block event channel on success.
-func (f *BlockFetcher) SubscribeNewBlockEvent(ctx context.Context) (<-chan types.EventDataSignedBlock, error) {
-	if f.isListeningForBlocks.Load() {
-		return nil, fmt.Errorf("already subscribed to new blocks")
-	}
-	ctx, cancel := context.WithCancel(ctx)
-	f.cancel = cancel
-	f.doneCh = make(chan struct{})
-	f.isListeningForBlocks.Store(true)
+func (f *BlockFetcher) runSubscriber() (chan types.EventDataSignedBlock, error) {
+	var (
+		subscription coregrpc.BlockAPI_SubscribeNewHeightsClient
+		err          error
+	)
 
-	subscription, err := f.client.SubscribeNewHeights(ctx, &coregrpc.SubscribeNewHeightsRequest{})
-	if err != nil {
-		return nil, err
-	}
+	signedBlockCh := make(chan types.EventDataSignedBlock, 1)
 
-	signedBlockCh := make(chan types.EventDataSignedBlock)
 	go func() {
 		defer close(f.doneCh)
 		defer close(signedBlockCh)
-		defer func() { f.isListeningForBlocks.Store(false) }()
+		timeout := time.NewTimer(retrySubscriptionDelay)
+		defer timeout.Stop()
 		for {
 			select {
-			case <-ctx.Done():
+			case <-f.ctx.Done():
+				return
+			default:
+			}
+
+			subscription, err = f.client.SubscribeNewHeights(f.ctx, &coregrpc.SubscribeNewHeightsRequest{})
+			switch {
+			case err == nil:
```

Comment:
grpc dialer will use exponential backoff retry by default. what is the point of timeout timer here? I think it can be removed and the code will be much simpler

Reply:
I guess backoff related to the grcp.Dial and not to the calling client's method.(removed timeout and starts spamming)

Reply:
Timer removed